swift/WorkflowUI/Sources/ViewControllerDescription/ViewControllerDescription.swift

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if canImport(UIKit)
+
+import UIKit
+
+
+/// A ViewControllerDescription acts as a recipe for building and updating a
+/// specific UIViewController.
+public struct ViewControllerDescription {
+
+    private let viewControllerType: UIViewController.Type
+    private let build: () -> UIViewController
+    private let update: (UIViewController) -> Void
+
+    /// Constructs a view controller description by providing a closure used to
+    /// build and update a specific view controller type.
+    ///
+    /// - Parameters:
+    ///   - type: The type of view controller produced by this description.
+    ///           Typically, should should be able to omit this parameter, but
+    ///           in cases where type inference has trouble, it’s offered as
+    ///           an escape hatch.
+    ///   - build: Closure that produces a new instance of the view controller
+    ///   - update: Closure that updates the given view controller
+    public init<VC: UIViewController>(type: VC.Type = VC.self, build: @escaping () -> VC, update: @escaping (VC) -> Void) {
+        self.viewControllerType = type
```

Comment:
Instead of doing this, what if we hold-on to the `canUpdate` block?

```
self.canUpdate = { untypedVC in
  return untypedVC is VC
}
```

Reply:
I played with that locally, but found it really useful for debugging to have the type around in a printable manner. I also don’t think we want `is` because we only want exact matches, not subclasses IMO.

Reply:
Since a single view controller description is responsible for building and updating, I can’t imagine a scenario where you’d reasonably build a type of a superclass and then be passed a value of a subclass, except by mistake.

Reply:
Do we actually want to be able to pass in something other than the default (`VC.self`) here?

Reply:
In the case of the extension on `ScreenViewController`, it was inferring a type of `ScreenViewController<FooScreen>` instead of `FooViewController` with older Swift versions (where I couldn’t do `Self(screen: screen)` and instead had to do `self.init(screen: screen)`. Adding this hinting allowed me to fix that.